File: millipede-test/src/test/java/millipede/ntrip/BaseNtrip1Emitter.java
package millipede.ntrip;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.Socket;

public class BaseNtrip1Emitter implements BaseNtripEmitter {
    private final String eol;
    private final String base;
    private final String password;
    private final Socket socket;

    public BaseNtrip1Emitter(String base, String password, String eol){
        this.base=base;
        this.password=password;
        this.eol=eol;

        socket=initializeSocket();
        try {
            sendMessage();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void stop(){
        try {
            socket.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private Socket initializeSocket(){
        try {
            Socket sock = new Socket("localhost", 2101);
            String request = """
                    SOURCE $passwd $base
                    Source-Agent: NTRIP LOADTEST/0.0.0
                    STR:
                    
                    """.replace("$base", base)
                    .replace("$passwd", password)
                    .replace("\n", eol);

            var os = sock.getOutputStream();
            os.write(request.getBytes());
            var in = new BufferedReader(new InputStreamReader(sock.getInputStream()));
            String s = "";
            while (!(s = in.readLine()).isEmpty()) {
                System.out.println(s);
            }
            System.out.println("Source " + base + " ok");
            return sock;
        }catch(Exception e){
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public void sendMessage() throws IOException {
        if (socket!=null){
            var os=socket.getOutputStream();
            os.write(("base ~" + base + "~\n").getBytes());
            os.flush();
        }
    }
}

File: caster-test/steps/utils/BaseNtrip1Emitter.py
import socket

from steps.utils.BaseNtripEmitter import BaseNtripEmitter


class BaseNtrip1Emitter(BaseNtripEmitter):
    def __init__(self, base, password, eol):
        self.base = base
        self.password = password
        self.eol = eol
        self.socket = self.initialize_socket()
        try:
            self.send_message()
        except Exception as e:
            print(e)

    def stop(self):
        try:
            self.socket.close()
        except Exception as e:
            print(e)

    def initialize_socket(self):
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.connect(("localhost", 2101))
            request = (
                f"SOURCE {self.password} {self.base}\n"
                f"Source-Agent: NTRIP LOADTEST/0.0.0\n"
                f"STR:\n"
                "\n"
            ).replace("\n", self.eol)
            sock.sendall(request.encode())
            in_stream = sock.makefile('r')
            s = ""
            while (s := in_stream.readline().strip()):
                print(s)
            print(f"Source {self.base} ok")
            return sock
        except Exception as e:
            print(e)
            return None

    def send_message(self):
        if self.socket:
            os = self.socket.makefile('w')
            os.write(f"base ~{self.base}~\n")
            os.flush()